File: 백준/Gold/1135. 뉴스 전하기/뉴스 전하기.java
/**
 * Author: nowalex322, Kim HyeonJae
 */

import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static int N, parent[], maxP, time[];
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

    public void solution() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
//        br = new BufferedReader(new InputStreamReader(new FileInputStream("src/main/java/BOJ_1135_뉴스전하기/input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        N = Integer.parseInt(br.readLine());
        parent = new int[N];
        time = new int[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(st.nextToken());
            parent[i] = num;
            maxP = num;
        }

        // 자식부터 시간 업데이트
        for(int i=N-1; i>=0; i--){
            updateTime(i);
        }

        bw.write(String.valueOf(time[0]));
        bw.flush();
        bw.close();
        br.close();
    }

    private void updateTime(int num) {
        ArrayList<Integer> childList = new ArrayList<>();
        for(int i=0; i<N; i++){
            if(parent[i] == num) childList.add(time[i]);
        }

        Collections.sort(childList, Collections.reverseOrder());

        int maxT = 0;
        for(int order = 1; order<=childList.size(); order++){
            maxT = Math.max(maxT, order + childList.get(order-1));
        }
        time[num] = maxT;
    }
}

File: 백준/Gold/1135. 뉴스 전하기/뉴스 전하기.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

int N;
vector<int> parent;
vector<int> childTime;
int maxT = 0;

void updateTime(int num) {
    vector<int> childList;
    for (int i = 0; i < N; i++) {
        if (parent[i] == num) childList.push_back(childTime[i]);
    }

    sort(ALL(childList), greater<int>());

    maxT = 0;
    for (int order = 1; order <= childList.size(); order++) {
        maxT = max(maxT, order + childList[order - 1]);
    }
    childTime[num] = maxT;
}

void solve() {
    cin >> N;
    parent.resize(N);
    childTime.resize(N);

    for (int i = 0; i < N; i++) {
        cin >> parent[i];
    }

    for (int i = N - 1; i >= 0; i--) {
        updateTime(i);
    }
    cout << childTime[0] << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}